### DISCUSSION TITLE
pool of ideas

### DISCUSSION THREAD

**Discussion by MoorPoor** [brunoherbelin/vimix#125] [Ideas]
hei!
I thought back and forth and came to the conclusion that a pool of ideas is the best for me...  and you then decide what you take seriously and what you don't, what is worth a separate issue or larger discussion and what is not. I'm sorry if I'm doing something wrong with this. deleting is always an option ;)

**Comment by MoorPoor** [comment]
**more free gripper distortion in F2**
what would you think of additional freely positionable grippers in order to be able to warp the sources within the area in F2 view?
once again very helpful for mapping matters
![photo_2024-02-19_21-06-34](https://github.com/brunoherbelin/vimix/assets/60827766/4e4dfc97-5b2c-44da-b216-cc827348839e)

**Reply by brunoherbelin** [reply]
Can work on this: behind the scene, now the texture of the rendering is projected on a high resolution mesh and non-linear interpolation and attractors can be applied to every vertex of the mesh: the problem is indeed more how to have a useful user interface... I didn't like to display the whole grid, as it is not very clear; but as you suggest could work: pick a point and move it... To be tested.

**Reply by MoorPoor** [reply]
yes, I think a small extra square will be enough nothing more. 
same to prevent certain parts to be effected -> add a point there ?!

**Comment by MoorPoor** [comment]
**separate standalone windows in F5 (and therefore in F2)**

I had a little trouble with using two projectors with one computer.

I end up creating a fairly broad project that is then stretched across both displays in F5. That worked semi-well so far because of different resolutions, but ok.

The add + window function gave me a brief hope that I could open an independent second window that I could then edit separately (or together) with the other one, but as far as I can see all I can do with it is duplicate the output window, right?

That's how I would imagine it:

![Bildschirmfoto vom 2024-02-19 21-19-14](https://github.com/brunoherbelin/vimix/assets/60827766/84121cda-b6aa-4dc0-ad80-02a233b76f99)

This would of course result in the need to be able to edit the new view... ideally, of course, as an additional window in F2, which can be optimally arranged freely. So horizontally, vertically, or even diagonally rotated :O. just like projectors are arranged in real life.

![Bildschirmfoto vom 2024-02-19 21-29-34](https://github.com/brunoherbelin/vimix/assets/60827766/d05b6014-59b7-403e-bf4f-e83425047d48)

**Reply by brunoherbelin** [reply]
You should be able to spread a unique (large) output framebuffer to two windows. What you have to do extra from what I see is to change the output mapping to the window (Custom Fit on the output in F5) and to disable the window title (Borderless from the menu of the output window in F5). Example :

![20240219212511959_vmixcapture](https://github.com/brunoherbelin/vimix/assets/60393189/936252ef-bd6a-4ed6-9618-c9a9573a6954)

This way each windows covers half of the output (horitontally). NB you can set the output framebuffer resolution to a custom one, e.g. two times 1920x1080 horizontally : 

![20240219212534320_vmixcapture](https://github.com/brunoherbelin/vimix/assets/60393189/38dd58a8-37d0-45c2-bda7-608efab18746)

Of note, if you have two full screen projections, you can assign one window 'Fullscreen on display 1', and the other window 'Fullscreen on display 2'.

**Reply by brunoherbelin** [reply]
Aha, and  I forgot to mention: to be precise in the scaling of the output mapping in the window, you can keep the `Alt` key down to activate the Snap mouse cursor to grid :

![20240219213252943_vmixcapture](https://github.com/brunoherbelin/vimix/assets/60393189/e9a0f2f0-0d94-4acf-a5a9-15c875e048de)

**Reply by MoorPoor** [reply]
I see that this is an option.

Regardless of the effort, I dream about how intuitive and great it would be if I press the little + and get another independent window. or a duplicate like now (certainly useful in some cases).

The F2 view is already designed as an “open world”, so additional windows would be wonderful to add, like in the fake screenshot above. 

again, I will not underestimate the fact that this would be a major intervention in the software.

**Reply by brunoherbelin** [reply]
I also like the 'open world' in F2: I might even find a drawing in my early notes that looks like yours :smile: 

In depth, it means that instead of one rendering output, vimix would have more than one (and F2 and the output preview would allow to show multiple). In practice, it is totally feasible. I will investigate how much deep it gets...

This doesnt however replace the F5 Displays view which is about how to split the rendering output in the computer monitors and do the projection corrections.

**Reply by MoorPoor** [reply]
sure, F5 remains important for distributing the outputs. I also like the projection correction.
I would also find it very practical if I could set the monitor/projector sizes in F5.
In practical use, you often have projectors with different orientations and different resolutions and aspect ratios.

**Reply by MoorPoor** [reply]
hello!
Given the circumstances, I'm coming back to the topic.
The option you suggested of working on multiple displays with independent footage is possible, but to be honest, it's really not convenient, especially with 3 or 4 projectors. it will be very hard for someone new without knowing how exactly to arrange that.
and that's a pity, because it is extremely powerful feature!

I also see a problem in the fact that once a project has been created, the size W&H can no longer be changed. It would be extremely practical if there was an option to change the resolution directly in the F5 view. Of course, it would be best if this were automatically adopted from the sum of all distributed windows.

I think this remains a very hot topic for many! At least for me :P

**Reply by brunoherbelin** [reply]
Hey! Thanks for keeping track of this hot topic :smile:  !

There are fundamental changes behind having multiple output rendering areas, which is why it was never addressed. I am however now ready to work on this. 

The main constraint is that a session cannot be limited by one display setup, i.e it should load and play on any computer disregarding the windows and  config of monitors in F5. Users should be able to load a session and be able to project it at different venues, adapting to the projectors and monitors available but not defining it based on the setup.

This is why the change of resolution of the session is in Session properties panel : 
(NB: I agree that the location of this parameter is hidden, and I will work on improving the UI.)
![image](https://github.com/user-attachments/assets/3265a092-b12b-4667-91bf-fce42343f93b)

In addition, the rendering of a session (i.e. loop over all sources) should be done only once for efficiency reasons, and not duplicated for each window / rendering area. And also, the display of output (tools/Display which also allows recording) should be independent of the projection windows  

Therefore, what I am thinking of implementing is the following:
  1. A session should only specify its render buffer resolution (as now)
  2. A session could have multiple rendering canvases, i.e. rectangular selections inside the global rendering buffer. This would be configured in Geometry view F2 (new)
  3. The dispatch of rendering canvases onto windows and monitors could be done in the Displays view F5 (new)

NB: step 3 is saved in the vimix settings on a local machine, not in the session. 
Would you think this is ok for your needs?

**Reply by MoorPoor** [reply]
As far as as I understand it sounds promising!

Do I understand correctly that there will be a large rendering buffer... basically a large work surface in which the individual screens of monitors or projectors will be placed?
Can the resolution of the global buffer then adjust automatically depending on how many canvases one like to put?

Thanks, that this load of work is not shocking you

**Comment by MoorPoor** [comment]
**ouput window autostart**
Another point that seems important to me is to have the option in the menu (under View?) as to how VIMIX autostarts? with the GUI or the output window in the foreground. essential for exhibition situations with only one screen.

**Reply by brunoherbelin** [reply]
I would need a bit more specifics to fully understand what you mean..
Do you think of a mode where there is no GUI window?

**Reply by MoorPoor** [reply]
well I like WYSIWYG visual programs where you can directly work in e.g. the projection without further gui > the request for source transforming in output window...
but here I simply want to point out, that as soon as you start vimix, the Gui is always in the foreground and the output window is in the background. For example, if I have vimix in autostart at an exhibition with a projector, I am forced to bring the output window to the foreground every time. Accordingly, permanent installation is practically not possible. UNLESS of course I have a control monitor connected.

**Reply by brunoherbelin** [reply]
I see and I also noticed issue #126 : I will revise the management of windows ASAP.

Would an 'Always on top' flag be useful ? (i.e. [floating window in GLFW](https://www.glfw.org/docs/3.3/window_guide.html#GLFW_FLOATING_attrib))

**Reply by lupin3rd** [reply]
Yes. I think that "always on top" can be a good solution in situations with only one screen. We need also a shortkey to set and unset this functionality.
In this way you can set correctly all setting and finally set "always on top".
But it is essential that the software remembers the last setting and automatically whatever happens (a restart of Vimix, a change of secondary monitor, ...) it resets the screen as it was last set (same position, width, height, ...).

**Reply by brunoherbelin** [reply]
Please try running **vimix** from command line with `--headless` (since Beta e3b8ccff9e4b29178569c6600255ab7a0eeab283)
This should execute **vimix** in the last executed state, but without the GUI (i.e. only the output windows). i.e. :

    vimix --headless

in command line.

I didn't do extensive testing, but my simple hack seems to work... Thanks for letting me know.

**Reply by brunoherbelin** [reply]
What I understand from your needs is a way to be 100% sure to enforce a starting status, whatever was done at one point. 

How I suggest to do this is to offer in vimix an 'Export settings' menu entry, that offers to save a vimix-config.xml, 
combined with a way to launch vimix with this settings file, e.g.

    vimix --settings /home/username/vimix-config.xml 

This way, you can always do whatever you want when you test vimix, and be sure the saved settings is restore at next startup using the config file. Would that solve what you need?

**Reply by brunoherbelin** [reply]
I implemented the `--settings` cmd-line option and added an 'Export settings' button in the settings panel of vimix. This seems to work fine. For example, first configure the output windows in fullscreen and export settings, and then launch :

    ./vimix --headless --settings /home/bh/fullscreen.xml
 
This will run vimix without GUI with the outputs in fullscreen as specified in the XML config.

**Reply by MoorPoor** [reply]
> Please try running vimix from command line with --headless (since Beta https://github.com/brunoherbelin/vimix/commit/e3b8ccff9e4b29178569c6600255ab7a0eeab283)

finally I managed to install beta again, had to get some more dependencies  sorted.
headless works good sofar... still need to get it running after boot

**Comment by MoorPoor** [comment]
**fullscreen also from menü?**
while in the menu, I would think it would make sense if the full screen function could also be found there... not particularly important for me, but I had to have a discussion with other users about why it can't be found ;)
![photo_2024-02-19_21-50-22](https://github.com/brunoherbelin/vimix/assets/60827766/0e657ab9-68a1-4a54-a78b-677e34c31ad6)

**Reply by brunoherbelin** [reply]
Thanks! Done Beta c1aa3c9d4d8fe1c65212e7a27e088b6a7518332a

**Comment by MoorPoor** [comment]
**Coffee**
ah, one more question,
Is there an option to send some money / or coffee to you here via Github?
In case someone makes money from the software via exhibitions, it could be a nice gesture to give something back... just assuming.

**Reply by brunoherbelin** [reply]
Thanks for the proposition!
I might indeed add a link inviting to support my friend artist Jean : https://vimeo.com/jeandetheux

**Comment by MoorPoor** [comment]
**one timeline for everyone**
soon I'll finish! what about if every source - no matter what it is - like you did with pictures.
I recently thought about that with a live stream that i wanted to fade in and out from time to time.

In this context, one could also reconsider artificially extending the timeline beyond the actual clip duration. The clip could continue to loop until the end of the timeline. This would give you more play in fading and other timeline actions

maybe also helpful in combination with chataigne

**Reply by brunoherbelin** [reply]
Well, this is where I have another idea that would do what you need, and more.

I want to add 'engines' in the GUI, i.e. automatons that you place in the workspace to perform a mechanical movements (e.g. move along a line in loops) : by attaching a source, it will follow the movement and thereby do a fade in and out, at the speed of the automaton. Speed curves could be defined to give the timeline and the automatons could be synchronized (reach this point triggers start of another automaton).

![ides_machine_vmixcapture](https://github.com/brunoherbelin/vimix/assets/60393189/023047e0-1ba5-4c47-9a3c-ddd052cb2ca8)

In short, the user would put some sort of machinery that moves the sources in the workspace, and this way define the succession of events, in loops or not. And all very visual and playful.. maybe too crazy?

**Reply by brunoherbelin** [reply]
Another idea, less visual, is simple scripting. I looked at https://chuck.stanford.edu/ as it is a time-based scripting. 
To be discussed....

**Reply by lupin3rd** [reply]
currently I do all these automations with Chataigne by OSC... interactions with the music and lights are really beautiful!
also a time-based scripting can be very interesting...

**Reply by MoorPoor** [reply]
Yes. chataigne gives a big boost of interaction options with sensors, mic, random commands etc... I still miss the camera blobtracker input ;) 
but it would be nice to have some automation or more advanced control options in vimix itself.
specially for people who are not that advanced in using software.

**Reply by MoorPoor** [reply]
> In short, the user would put some sort of machinery that moves the sources in the workspace, and this way define the succession of events, in loops or not. And all very visual and playful.. maybe too crazy?

can't be crazy enough!

**Comment by lupin3rd** [comment]
**not only current, also next**
There are many situations where mixing occurs between two videos. very often I use midi controllers with sliders to mix videos. currently the only possibility of selection is using the "current" or specifying the id or name of the video, but in this way we have to assign a slider to each video.
If we had another selection in addition to "current", for example "next" we could use them to carry out these mixes or even to decide how to move from current to next.

**Reply by brunoherbelin** [reply]
Doesn't `/vimix/current/next` and `/vimix/current/previous` work ? Or do you mean something else ?

**Reply by lupin3rd** [reply]
I meant another pointer would be useful. Two "pointer": current and next

**Reply by brunoherbelin** [reply]
aha, what you mean is to have variables :) Or fixed keywords that refer to different sources depending on the situation or the session.

One option could be to target a 'batch' of sources (which can be a single source) :
https://github.com/brunoherbelin/vimix/wiki/Open-Sound-Control-API#target-a-batch-of-sources 
Your GUI could be made for batch#0 and batch#1, disregarding which source is in the batches?

What also could work is to use an OSC translator : depending on the translation rules (which can be changed), a given keyword of your choice (e.g. `mytarget`) can be associated to source number `#1` or `#2`, etc.. The translator is not very much used (it was for accepting OSC input from fixed-syntax apps), but I could add some dynamic configuration rules that a client defines. E.g. could inform vimix that, from now on, when it receives target `toto`, it means `#1`. Then later on, re-assign `toto` to `#2`. etc.

What would you prefer / suggest ?

**Reply by brunoherbelin** [reply]
I have an idea, and I think you will like it. I will add `alias` to OSC:

1. send to vimix the alias to use for a given source


       /vimix/#2/alias s whatiwant


2. use the alias to name the source : 


       /vimix/whatiwant/alpha f 0.2


=> this will target the source `#2` via the alias `whatiwant`

**Reply by lupin3rd** [reply]
This is a great idea because it gives many possibilities and solves a lot of situations.

**Reply by brunoherbelin** [reply]
Implemented in Beta f280d3b64cb7b7eeed1524a67afea0461a21f1e4. Just tested with this:

        oscsend localhost 7000 /vimix/#0/alias s toto
        oscsend localhost 7000 /vimix/toto/alpha f 0

and it seems to work fine. Importantly the aliases are temporary, working on the ongoing vimix.

I also added a way to rename a source, which in this case is saved in the session (as normally expected for the name of the source.

        oscsend localhost 7000 /vimix/#0/rename s toto
        oscsend localhost 7000 /vimix/toto/alpha f 0

Note that with rename, the name and initials of the source change in the GUI too.

**Reply by gleb812** [reply]
cool! I like this feature! should be helpful in some control cases
